iree/compiler/Dialect/HAL/Transforms/MaterializeInterfaces2.cpp

```diff
@@ -260,47 +92,1009 @@ static LogicalResult declareTargetOps(TargetOptions targetOptions,
+  OperandResultRef(int type, int value) : type(type), value(value) {}
+
+  bool isOperand() const { return type == 0; }
+  bool isResult() const { return type == 1; }
+  int index() const { return value; }
+
+ private:
+  int type : 1;
+  int value : 31;
+};
+
+// HACK: only needed because we don't have the right IR ops to represent this.
+// I hate this.
+static OperandResultRef mapRegionOperandToDispatchValue(
+    FuncOp entryFuncOp, int regionOperandIndex) {
+  auto types = entryFuncOp.getType().getInputs();
+  int operandIndex = 0;
+  int resultIndex = 0;
+  for (int i = 0; i < regionOperandIndex; ++i) {
+    auto operandType = types[i];
+    if (operandType.isIntOrIndexOrFloat()) {
+      ++operandIndex;
+    } else if (auto tensorType =
+                   operandType.dyn_cast<IREE::Flow::DispatchTensorType>()) {
+      switch (tensorType.getAccess()) {
+        case IREE::Flow::TensorAccess::ReadOnly:
+          ++operandIndex;
+          break;
+        case IREE::Flow::TensorAccess::ReadWrite:
+          ++operandIndex;
+          ++resultIndex;
+          break;
+        case IREE::Flow::TensorAccess::WriteOnly:
+          ++resultIndex;
+          break;
+      }
+    }
+  }
+  auto operandType = types[regionOperandIndex];
+  if (operandType.isIntOrIndexOrFloat()) {
+    return OperandResultRef::makeOperand(operandIndex);
+  } else if (auto tensorType =
+                 operandType.dyn_cast<IREE::Flow::DispatchTensorType>()) {
+    switch (tensorType.getAccess()) {
+      case IREE::Flow::TensorAccess::ReadOnly:
+        return OperandResultRef::makeOperand(operandIndex);
+      case IREE::Flow::TensorAccess::ReadWrite:
+        return OperandResultRef::makeOperand(operandIndex);
+      case IREE::Flow::TensorAccess::WriteOnly:
+        return OperandResultRef::makeResult(resultIndex);
+    }
+  }
+  llvm_unreachable("invalid region operand/result mapping");
+}
+
+// Information about dispatch region operand usage derived from flow.dispatches.
+struct DerivedOperandInfo {
+  enum Usage : unsigned {
+    // Unused or not yet scanned.
+    UNKNOWN = 0,
+    // I/O from a constant pool storage buffer.
+    CONSTANT_BUFFER = 1u << 0,
+    // I/O to a transient storage buffer.
+    TRANSIENT_BUFFER = 1u << 1,
+    // I/O to an external/unknowable buffer (invocation capture/escape).
+    EXTERNAL_BUFFER = 1u << 2,
+    // Push constant (index) non-buffer type.
+    PUSH_CONSTANT = 1u << 3,
+  };
+
+  // Operand index in the flattened dispatch function.
+  int operandIndex = 0;
+
+  // Bitfield of usage throughout the program.
+  // Entry points dispatched from many places may have differing usage.
+  unsigned usage = Usage::UNKNOWN;
+
+  // Storage buffer and ranges pulled from it if this has CONSTANT_BUFFER usage
+  // and all uses are of the same storage buffer. Unset if multiple storage
+  // buffers are used.
+  SymbolRefAttr constantBuffer;
+  SmallVector<Attribute> constantRanges;
+
+  void dump() {
```

Comment:
This debug output is great for inspecting this pass, but it's also very verbose (judging by your examples for complex models). Is there guidance from LLVM/MLIR on how to use `llvm:dbgs()` like this? Should this be further flag guarded, or is the idea that people turning on debug logging would be targeted with which passes they are running?

Reply:
I debated about that with myself as well. I settled on leaving them in as when we started using linalg our default `-debug` output became useless as it spams a tremendous amount and `-debug-only=` is the only real way to get usable results - in which case, the logs in this file only show with `-debug-only=iree-hal-materialize-interfaces`. There's not a way (that I could find) to only enable these if and only if -debug-only= is set - LLVM_DEBUG will either always log if -debug is passed or only log if -debug-only= is passed with the right name.

Since this is one of those areas that I expect compilation errors to happen in I'm going to leave them in for now so that it's easy to get reports from people (no need to go change code, just pass `-debug-only=iree-hal-materialize-interfaces` and pipe to a file).